### PR TITLE
[LA-2314] reusable workflow for format& run unit test for golang projects

### PR DESCRIPTION
**Pull request by FatmaTosun**
## What does this change?

create reusable workflow for format and run unit test for golang projects

## Why?

We want a shared github workflow that handles formatting and testing of our go projects, so that all our projects are more similar.

## Link to supporting ticket or Screenshots (if applicable)
https://axelspringer.atlassian.net/browse/LA-2314

### CODE REVIEW

File: .github/workflows/reusable-workflow__golang__format-unit-test.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    secrets:
+      ACCESS_KEY_ID:
+        required: true
+      ACCESS_KEY_SECRET:
+        required: true
+      LA_TECH_USER_AUTH_TOKEN:
+        required: true
+      LA_TECH_USER_SSH_KEY:
+        required: true
+      GO_VERSION:
```
Comment (FatmaTosun) [inline .github/workflows/reusable-workflow__golang__format-unit-test.yml:14]: I prefer to pass `GO_VERSION` in case we want to upgrade version in our projects.

File: .github/workflows/reusable-workflow__golang__format-unit-tests.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      ACCESS_KEY_ID:
+        required: true
+      ACCESS_KEY_SECRET:
+        required: true
+      LA_TECH_USER_AUTH_TOKEN:
+        required: true
+      LA_TECH_USER_SSH_KEY:
+        required: true
+      GO_VERSION:
+        required: true
```
Comment (k-pusch) [inline .github/workflows/reusable-workflow__golang__format-unit-tests.yml:15]: Do we have a "standard" default go version we could use as a default? and just overwrite it when we need a specific go version?

File: README.md [hunk 3]
```diff
@@ -165,6 +166,25 @@ jobs:
...
+jobs:
+  …
+  format-and-unit-tests:
+    uses: spring-media/la-shared-github-workflows/.github/workflows/reusable-workflow__golang__format-unit-tests.yml@v1
+    secrets:
+      ACCESS_KEY_ID: ${{ secrets.access_key_id }}
+      ACCESS_KEY_SECRET: ${{ secrets.access_key_secret }}
+      LA_TECH_USER_AUTH_TOKEN: ${{ secrets.LA_TECH_USER_AUTH_TOKEN }}
+      LA_TECH_USER_SSH_KEY: ${{ secrets.la_tech_user_ssh_key }}
+      GOLANG_VERSION: ${{ secrets.GO_VERSION }}
```
Comment (k-pusch) [inline README.md:185]: I am not sure if we want to have the GO_VERSION as a github secret, or? Usually we set all the secrets in la-hollywood and those are mostly just keys and passwords. Why not passing it in as a string/number value? 

Or am i missing something? Just wondering if this documentation would lead to confusion.

**Review by k-pusch — COMMENTED** [review]
I have some small questions. Otherwise it looks pretty straight forward to me.

File: .github/workflows/reusable-workflow__golang__format-unit-tests.yml (hunk 2 above)
Comment (nhoffmann) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:15]: I would always be explicit here. Less magic.

File: README.md (hunk 3 above)
Comment (nhoffmann) [reply README.md:185]: True. It is misleading. It is highly unlikely the version will come from a Github secret.

Comment (nhoffmann) [reply README.md:185]: I think we can make this more clear by making the GOLANG_VERSION an input to the shared workflow instead of a secret. It is technically the same, but semantically more clear.

https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow

File: .github/workflows/reusable-workflow__golang__format-unit-tests.yml (hunk 2 above)
Comment (k-pusch) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:15]: Fair point

File: README.md (hunk 3 above)
Comment (FatmaTosun) [reply README.md:185]: true, I will pass it as input.

Comment (FatmaTosun) [reply README.md:185]: done.

Comment (nhoffmann) [reply README.md:185]: The readme needs adjusting too 😇

File: .github/workflows/reusable-workflow__golang__format-unit-tests.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+          touch ~/.ssh/config
+          export location=~
+          cat << EOF > ~/.ssh/config
+            Host github.com
+              User git
+              Hostname github.com
+              PreferredAuthentications publickey
+              IdentityFile $location/.ssh/id_rsa
+          EOF
+          git config --global url."https://${{ secrets.LA_TECH_USER_AUTH_TOKEN }}:x-oauth-basic@github.com/spring-media".insteadOf "https://github.com/spring-media"
```
Comment (nhoffmann) [inline .github/workflows/reusable-workflow__golang__format-unit-tests.yml:75]: Do we really need this step for testing and formatting? If not, can we remove it? Or am I missing something?

File: .github/workflows/reusable-workflow__golang__format-unit-tests.yml [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          aws-access-key-id: ${{ secrets.ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.ACCESS_KEY_SECRET }}
+          aws-region: eu-central-1
+
+      - name: Install Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ inputs.go_version }}
+
+      - name: Check go modules cache
```
Comment (nhoffmann) [inline .github/workflows/reusable-workflow__golang__format-unit-tests.yml:39]: I think we want to adjust the step name. The name here is similar to the name of the following step that verifies whether everything is properly formatted. 
But this step here is still setup, it tries to restore go modules from a cache to speed up the setup.

I think it will avoid confusion if we make it more clear, that this is still setup and not actually running a check on the codebase.

File: README.md [hunk 6]
```diff
@@ -165,6 +166,27 @@ jobs:
   …
 ```
 
+## golang\_\_format-unit-tests
+
+The [golang\_\_format-unit-tests workflow](./.github/workflows/reusable-workflow__golang__format-unit-tests.yml) will set up golang version and format & run unit tests on the project:
```
Comment (nhoffmann) [inline README.md:171]: Maybe nitpicky, but I would make it clear that it does not format anything, it checks whether everything is formatted properly and errors otherwise, right?

**Review by nhoffmann — CHANGES_REQUESTED** [review]
Some suggestions on how to make this even better and clearer.

File: README.md (hunk 6 above)
Comment (FatmaTosun) [reply README.md:171]: yes, that's true. I will make it clearer `will set up golang version and check formatted properly and run unit tests on the project `

File: .github/workflows/reusable-workflow__golang__format-unit-tests.yml (hunk 4 above)
Comment (FatmaTosun) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:75]: we need pull dependencies we can't remove it.

Comment (FatmaTosun) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:75]: It's not related being shared-workflow or not we need it for running unit tests. Our unit tests are not completely mocked so we need to pull dependencies. For example;

https://github.com/spring-media/la-hua-api/blob/main/internal/api/utils_unit_test.go#L23

we do call `profile.NewMockedUserProfileClient` so we need to have `"github.com/spring-media/la-golang-shared/profile"` . If it's not already cached we have to pull it.

Comment (nhoffmann) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:75]: Ok, so it's not a direct dependency. We only need it in some projects because when running tests, we might need to pull in dependencies that are only available in private Github repositories. 

Since that all is not obvious from looking at the workflow, can we put this information into the shared workflow? As a comment maybe?

Comment (FatmaTosun) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:75]: sure, I will add comment.

Comment (FatmaTosun) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:77]: added.

Comment (nhoffmann) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:77]: Thanks! I think this will our future selfs in understanding this.

**Review by nhoffmann — APPROVED** [review]
Looks good to me. 
We can only know if everything is in order by running it though. This is Github Actions, something one always forgets. You could test it by pointing one of the PRs that are already open to start using it against the latest git commit of this branch. Or maybe you did that already.

**Comment by FatmaTosun** [comment]
> Looks good to me.
> We can only know if everything is in order by running it though. This is Github Actions, something one always forgets. You could test it by pointing one of the PRs that are already open to start using it against the latest git commit of this branch. Or maybe you did that already.

Actually, I didn't :/ After your comment I tested interactive-api against latest commit and I realised we can not reference secrets directly. https://github.com/spring-media/la-shared-github-workflows/pull/23/commits/9bd26fb1ee6fc7be81f709bd509cc814da195ed9 Thank you 🙏

File: .github/workflows/reusable-workflow__golang__format-unit-tests.yml [hunk 7]
```diff
@@ -76,6 +79,14 @@ jobs:
...
           git config --global url."https://${{ secrets.LA_TECH_USER_AUTH_TOKEN }}:x-oauth-basic@github.com/spring-media".insteadOf "https://github.com/spring-media"
 
+      # this step is needed for bouncer casino auth and can be removed after migrating to cognito auth
+      - name: Copy bouncer private key
+        env:
+          BOUNCER_PRIVATE_KEY: ${{ secrets.BOUNCER_PRIVATE_KEY }}
+        if: secrets.BOUNCER_PRIVATE_KEY != ""
+        run: |
+          echo "$BOUNCER_PRIVATE_KEY" > resources/bouncer_private.pem
+          chmod 600 resources/bouncer_private.pem
```
Comment (nhoffmann) [inline .github/workflows/reusable-workflow__golang__format-unit-tests.yml]: This introduces a bit too much magic for my taste. I think this should go into the workflow that needs the private key. The reason would be to make it more clear in the project where that pem is coming from, where it is defined. It is hard to follow when a shared workflow that is about. This also has nothing to do with linting and unit testing.

**Review by nhoffmann — CHANGES_REQUESTED** [review]
I am against pulling the bouncer key into this shared workflow. It is mixing concerns, which will lead to confusion.

File: .github/workflows/reusable-workflow__golang__format-unit-tests.yml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+      # this step is needed for bouncer casino auth and can be removed after migrating to cognito auth
+      - name: Copy bouncer private key
+        env:
+          BOUNCER_PRIVATE_KEY: ${{ secrets.BOUNCER_PRIVATE_KEY }}
+        run: |
+          if [ "$BOUNCER_PRIVATE_KEY" != '' ]; then
+            echo "$BOUNCER_PRIVATE_KEY" > resources/bouncer_private.pem
+            chmod 600 resources/bouncer_private.pem
+          fi
```
Comment (nhoffmann) [inline .github/workflows/reusable-workflow__golang__format-unit-tests.yml:90]: Can you elaborate why we need to have this in the shared workflow?
This introduces a bit too much magic for my taste. I really think this should go into the workflow that needs the private key. The reason would be to make it more clear in the project, where that pem is coming from, where it is defined. It is hard to follow when it is coming from a shared workflow that is about unit tests. This has nothing to do with linting and unit testing.

Comment (nhoffmann) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:90]: Do we need that bouncer_private.pem to be present to run the tests?

Comment (FatmaTosun) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:90]: This private key defined in secrets and it is needed for generating jwt token with casinoClient. We have 2 endpoints that generate code with this private key.

`/qrcode/validate`
`/qrcode/initValidation`

About why we need to have this in shared workflow, because we have unit test for `/qrcode/validate`(TestValidateEndpoint) and it's not completely mocked. But I can mock encodeCode part and remove this `Copy bouncer private key` step from here.

Comment (nhoffmann) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:90]: IIUC we only need this for la-bouncer-apis workflow. I was just wondering if we can move it to that project and not have it in the shared library. 

I would really prefer not using the shared workflow for bouncer-api then, instead of hacking something in here temporarily. We can start using the shared workflow for bouncer-api, when we removed casino from bouncer-api entirely.

Comment (FatmaTosun) [reply .github/workflows/reusable-workflow__golang__format-unit-tests.yml:90]: yes, that's correct we only need for la-bouncer-api. 

I will remove it from workflow then.

**Review by nhoffmann — APPROVED** [review]
Very nice!

**Comment by nhoffmann** [comment]
@FatmaTosun You probably know it but here's a reminder anyway: After merging the branch we need to release a new minor version and update the tag as described here https://github.com/spring-media/la-shared-github-workflows#releasing
